File: front/lib/models/messageModel.dart
import 'package:flutter/material.dart';
import 'userModel.dart';
enum MessageType { text, image, audio, file }

class Message {
  final int id;
  final String content;
  final String senderId;
  final String receiverId;
  final MessageType type;
  final String? fileUrl;
  final bool isRead;
  final DateTime createdAt;
  final DateTime updatedAt;
  final User? sender;
  final User? receiver;

  Message({
    this.id = 0,
    required this.content,
    required this.senderId,
    required this.receiverId,
    this.type = MessageType.text,
    this.fileUrl,
    this.isRead = false,
    required this.createdAt,
    required this.updatedAt,
    this.sender,
    this.receiver,
  });

  factory Message.fromJson(Map<String, dynamic> json) {
    return Message(
      id: json['id'] ?? 0,
      content: json['content'] ?? '',
      senderId: json['senderId'] ?? '',
      receiverId: json['receiverId'] ?? '',
      type: _getMessageType(json['type']),
      fileUrl: json['fileUrl'],
      isRead: json['isRead'] ?? false,
      createdAt: json['createdAt'] != null 
          ? DateTime.parse(json['createdAt']) 
          : DateTime.now(),
      updatedAt: json['updatedAt'] != null 
          ? DateTime.parse(json['updatedAt']) 
          : DateTime.now(),
      sender: json['sender'] != null ? User.fromJson(json['sender']) : null,
      receiver: json['receiver'] != null ? User.fromJson(json['receiver']) : null,
    );
  }

  static MessageType _getMessageType(String? type) {
    switch (type) {
      case 'image':
        return MessageType.image;
      case 'audio':
        return MessageType.audio;
      case 'file':
        return MessageType.file;
      default:
        return MessageType.text;
    }
  }

  Map<String, dynamic> toJson() {
    return {
      'content': content,
      'senderId': senderId,
      'receiverId': receiverId,
      'type': type.toString().split('.').last,
      if (fileUrl != null) 'fileUrl': fileUrl,
    };
  }
}

File: Backend_Project/models/messageModel.js
const Message = (sequelize, DataTypes) => {
  const Message = sequelize.define('Message', {
    id: {
      type: DataTypes.INTEGER,
      primaryKey: true,
      autoIncrement: true
    },
    senderId: {
    type: DataTypes.INTEGER,
    allowNull: false
  },
  receiverId: {
    type: DataTypes.INTEGER,
    allowNull: false
  },
    content: {
      type: DataTypes.TEXT,
      allowNull: false
    },
    type: {
      type: DataTypes.ENUM('text', 'image', 'audio', 'file'),
      defaultValue: 'text'
    },
    fileUrl: {
      type: DataTypes.STRING
    },
    isRead: {
      type: DataTypes.BOOLEAN,
      defaultValue: false
    }
  });

  Message.associate = (models) => {
    Message.belongsTo(models.User, {
      foreignKey: 'senderId',
      as: 'sender'
    });
    Message.belongsTo(models.User, {
      foreignKey: 'receiverId',
      as: 'receiver'
    });
  };

  return Message;
};

export default Message;